Update with more commands, some more functionality, some fixes

WolfoIsBestWolf opened the pull request:
Hope more people can get to use this stuff.

If some stuff needs to be changed let me know.


Only known issue; `no_cooldowns` doesn't work with SS2 due to some old code on their end, not really anything that can be done here.

WolfoIsBestWolf committed:
fixes / macros / personal commands ported

WolfoIsBestWolf committed:
probably finished

WolfoIsBestWolf committed:
send it

WolfoIsBestWolf committed:
Update Lang.cs

WolfoIsBestWolf committed:
Merge branch 'master' into master

SChinchi commented on Code/AutoCompletion/AutoCompleteManager.cs:
```diff
@@ -90,9 +91,22 @@ internal static void RegisterAutoCompleteCommands()
 
             parser.RegisterDynamicVariable("director_card", StringFinder.Instance.DirectorCards, "spawnCard", autocompleteIndex: 1);
-            parser.RegisterDynamicVariable("interactable", StringFinder.Instance.InteractableSpawnCards, autocompleteIndex: 1);
+            //parser.RegisterDynamicVariable("interactable", StringFinder.Instance.InteractableSpawnCards, autocompleteIndex: 1);
```
Commented out code should be removed unless it serves some specific purpose, e.g., commented out temporarily to disable functionality. Obsolete code just adds noise to the codebase and we also have the diffs to see what was there before in case we need to revert something.

SChinchi commented on Code/AutoCompletion/AutoCompleteManager.cs:
```diff
@@ -90,9 +91,22 @@ internal static void RegisterAutoCompleteCommands()
             parser.RegisterDynamicVariable("director_card", StringFinder.Instance.DirectorCards, "spawnCard", autocompleteIndex: 1);
-            parser.RegisterDynamicVariable("interactable", StringFinder.Instance.InteractableSpawnCards, autocompleteIndex: 1);
+            //parser.RegisterDynamicVariable("interactable", StringFinder.Instance.InteractableSpawnCards, autocompleteIndex: 1);
+            parser.RegisterDynamicVariable("interactable", StringFinder.Instance.InteractableSpawnCards.Select(i => $"{StringFinder.Instance.InteractableSpawnCards.IndexOf(i)}|{i.name}|{StringFinder.GetLangInvar(i.prefab?.GetComponent<IDisplayNameProvider>()?.GetDisplayName())}"), autocompleteIndex: 1);
```
Thank you for this addition, very useful. It seems that since all spawn cards are captured on load, this does not need to be a dynamic variable anymore. Static variables are for fixed catalogs, while dynamic variables are for iterables that can change content at runtime, e.g., player list.

If I've misunderstood something and this still needs to be a dynamic variable, remove the IndexOf from the string part, just keep the name and display name. Dynamic variables automatically prefix the index whenever requested and the way it's written right now we end up with "index|index|internalName|displayName".

SChinchi commented on Code/AutoCompletion/AutoCompleteManager.cs:
```diff
@@ -90,9 +91,22 @@ internal static void RegisterAutoCompleteCommands()
             parser.RegisterStaticVariable("permission_level", CollectEnumNames(typeof(Permissions.Level), typeof(int)), 1);
 
             parser.RegisterDynamicVariable("director_card", StringFinder.Instance.DirectorCards, "spawnCard", autocompleteIndex: 1);
```
Can we do the same for the csc as we have done for the isc? Related comment on line 95.

SChinchi commented on Code/DT-Commands/CurrentRun.cs:
```diff
@@ -681,6 +769,14 @@ private static void CCSetTime(ConCommandArgs args)
+        public static void CC_evolve_lemurian(ConCommandArgs args)
+        {
+            DevotionInventoryController.ActivateAllDevotedEvolution();
+        }
```
- I don't think we need an alias here as it polutes the command catalog. We generally use aliases for common mispellings or for a command that does multiple things (which this one doesn't). Here one is a prefix of the other; just pick either of the spellings and go with it.
- The helpText should be a constant in `Lang.cs`.
- Check for run instance and log if false. Server commands can run in the lobby, but most such commands only make sense within a run.
- Consider adding an optional argument "self|all" with the default being self. Evolving everyone's lemurians can be more destructive than intended. You can iterate the instances list and compare the controller's master to `args.senderMaster` to find your own.
- Nitpick but naming the method `CCEvolveLemurian` would be nicer for consistency.

SChinchi commented on Code/DT-Commands/Drones.cs:
```diff
@@ -0,0 +1,213 @@
+        [AutoComplete(Lang.GIVEDRONE_ARGS)]
+        private static void CCGiveDrone(ConCommandArgs args)
+        {
+            //give_drone {name} {amount} {tier}
```
Remove comment or correct it; the arguments are `name, amoumt, tier, target`. There is also a specific meaning to curly and square brackets as explained in the README. Personally I'd just remove the comment so we don't have to deal with stale comments should the arguments ever change. If one needs a refresher, `Lang.GIVEDRONE_ARGS` is just two lines above it for an xref.

SChinchi commented on Code/DT-Commands/Drones.cs:
```diff
@@ -0,0 +1,213 @@
+            int tier = 1;
+            if (args.Count > 2 && args[2] != Lang.DEFAULT_VALUE && !TextSerialization.TryParseInvariant(args[2], out tier))
+            {
+                Log.MessageNetworked(String.Format(Lang.PARSE_ERROR, "tier", "int"), args, LogLevel.MessageClientOnly);
```
Inconsistent variable name in the parsing error. Ideally this should match what is declared in the ARGS. For that matter I favour `drone_tier` for naming consistency.

SChinchi commented on Code/DT-Commands/Items.cs:
```diff
@@ -187,9 +227,121 @@ private static void CCGiveItem(ConCommandArgs args)
+        
+        }
+
+ 
```
Duplicated command; does not belong here. Ironically enough in this one (which overrides the one in Drones.cs) you do check for negative value arguments.

WolfoIsBestWolf committed:
set_crit & spawn_as permament1 default

WolfoIsBestWolf committed:
Merge branch 'master' of https://github.com/WolfoIsBestWolf/DebugToolkit

SChinchi commented on Code/DT-Commands/Items.cs:
```diff
@@ -824,7 +1060,14 @@ private static BasicPickupDropTable ParseDroptable(ConCommandArgs args, int inde
+                droptable.Add(availableDropLists[ItemTier.Tier2], 1f);
+                droptable.Add(availableDropLists[ItemTier.Tier3], 1f);
+                droptable.Add(availableDropLists[ItemTier.Boss], 1f);
+            }
```
This is more of a thought, but if we're going to have a limited tier list as the default, with which I agree since voids corrupting your normal items is very annoying, we should probably have custom weights for each tier, akin to `midgame` in Macros.cs. That way the user can get a nice item distribution by default.

SChinchi commented on Code/DT-Commands/Drones.cs:
```diff
@@ -0,0 +1,213 @@
+using UnityEngine.AddressableAssets;
+using UnityEngine.Networking;
+using static DebugToolkit.Log;
+using static DebugToolkit.Util;
```
The majority of these are unused; clean up please.

SChinchi commented on Code/DT-Commands/LobbyManagement.cs:
```diff
@@ -66,10 +66,23 @@ private static void CCTrueKill(ConCommandArgs args)
+            
+            }
+            Log.MessageNetworked("Everyone was killed by server.", args);
+        }
```
Unnecessary duplication. Just modify `Lang.TRUEKILL_ARGS` to `[player (player|'all'):<self>]` and handle the logic in the original `true_kill` command.

SChinchi commented on Code/DT-Commands/Macros.cs:
```diff
@@ -36,14 +37,50 @@ private static void LateGame(ConCommandArgs args)
+            Invoke(args.sender, "kill_all", "2", "1");
+            Invoke(args.sender, "kill_all", "4", "1");
+            Invoke(args.sender, "no_enemies", "1");
+            //Vanilla peace also sets god 1 but ehh idk i dont need that
```
Others might. Note that even though no enemies spawn, you are still be vulnerable to other damage sources, such as void fog.

SChinchi commented on Code/DT-Commands/Macros.cs:
```diff
@@ -36,14 +37,50 @@ private static void LateGame(ConCommandArgs args)
+            Invoke(args.sender, "give_item", "boostdamage", "9999990");
+        }
+
+        [ConCommand(commandName = "scanner", flags = ConVarFlags.ExecuteOnServer, helpText = Lang.MACRO_SCANNER_HELP)]
```
Similar to the "peace" vs "dtpeace" point.

SChinchi commented on Code/DT-Commands/Miscellaneous.cs:
```diff
@@ -79,5 +80,53 @@ static IEnumerator InvokeRoutine(System.Action action, float delay)
+            prevTimeScale = Time.timeScale;
+            Time.timeScale = newTimeScale;
+            TimescaleNet.Invoke(newTimeScale);
+        }
```
Why not in CurrentRun.cs?

SChinchi commented on Code/DT-Commands/PlayerCommands.cs:
```diff
@@ -49,14 +55,74 @@ private static void CCBuddhaModeToggle(ConCommandArgs args)
 
+        [ConCommand(commandName = "buddhaenemy", flags = ConVarFlags.ExecuteOnServer, helpText = Lang.BUDDHA_HELP2)]
+        [ConCommand(commandName = "budaenemy", flags = ConVarFlags.ExecuteOnServer, helpText = Lang.BUDDHA_HELP2)]
+        [ConCommand(commandName = "invulenemy", flags = ConVarFlags.ExecuteOnServer, helpText = Lang.BUDDHA_HELP2)]
```
Happy with the buddha spelling variations, but `invulenemy` is not very descriptive. Why don't we just add a second argument to `buddha`, i.e., `[target_players (0|1):1]`, and handle the logic there?

SChinchi commented on Code/DT-Commands/PlayerCommands.cs:
```diff
@@ -49,14 +55,74 @@ private static void CCBuddhaModeToggle(ConCommandArgs args)
+                    {
+                        body.bodyFlags &= ~RoR2.CharacterBody.BodyFlags.Buddha;
+                    }
+                }
```
If I'm reading this properly, this would also toggle it for your minions since they aren't players but the command description clearly states it's only monsters. This also affects pots and other destructibles, so a masterless check would make sense. Either that or clarify in the help text that any non-player body is affected.

SChinchi commented:
> Yeah there's quite a lot. Big theme is just ease-of-use hope that's appreciated.
> 
> I use stuff like `scanner` hundreds of times I assume people will like it if they find it.

It is indeed appreciated. I'm still going through the review and so far there only a few objections on either the implementation of something or whether something should be included. I expect the other members to offer their opinion if there is a disagreement, so don't take everything I say at face value. But brace yourself for A LOT of comments, mainly for code quality and some minor issues. I might leave some classes for a second round of reviewing just so I can get the rest to you in a prompt manner.

> I'll put "spawn_as" the default to permanent if this was seen as more of a bug fix anyways. Don't know if spawning as something temporarily would see much use but probably fine to have it as a an option.

Yep, it was just one-line bug fix. Just because an extra option is now available, it doesn't always mean it adds any real value if it's too niche. Scope creep vs simplicity/functionality and all.

SChinchi commented on Code/DT-Commands/PlayerCommands.cs:
```diff
@@ -223,6 +372,12 @@ private static void CCHurt(ConCommandArgs args)
+            bool bypassCalc = false;
+            if (args.Count > 2 && args[2] != Lang.DEFAULT_VALUE && !Util.TryParseBool(args[2], out bypassCalc))
+            {
+                Log.MessageNetworked(String.Format(Lang.PARSE_ERROR, "bypassCalc", "int or bool"), args, LogLevel.MessageClientOnly);
```
The standard is to just say "bool" here. Also could we rename this to something more descriptive? I don't mind a fully spelt out "bypass_calculations", though I seem to gravitate more towards "direct". Shorter names are generally more preferable to minimise mental parsing, but not at the expensive of clarity. Don't forget that the help text can also provide more insight on its usage.

SChinchi commented on Code/DT-Commands/PlayerCommands.cs:
```diff
@@ -606,5 +763,318 @@ internal static bool UpdateCurrentPlayerBody(out NetworkUser networkUser, out Ch
+       
+      
+        [ConCommand(commandName = "nocooldown", flags = ConVarFlags.ExecuteOnServer, helpText = Lang.NOCOOLDOWN_HELP)]
+        [ConCommand(commandName = "nocooldowns", flags = ConVarFlags.ExecuteOnServer, helpText = Lang.NOCOOLDOWN_HELP)]
```
Choose either or but not both.

SChinchi commented on Code/DT-Commands/PlayerCommands.cs:
```diff
@@ -606,5 +763,318 @@ internal static bool UpdateCurrentPlayerBody(out NetworkUser networkUser, out Ch
+            mdl.SetActive(!mdl.activeSelf);
+            Log.MessageNetworked(String.Format(!mdl.activeSelf ? Lang.SETTING_ENABLED : Lang.SETTING_DISABLED, "Invisible Model"), args);
+
+        }
```
- I like `hide_model` the best, the rest aren't worthy aliases as it stands.
- helpText Lang.cs constant
- CCToggleModel
- Run.instance and args.senderBody null checks (since you could be dead when calling it). The command can have None flags since you want its effect locally, but it only makes sense in the context of a run.

SChinchi commented on Code/DT-Commands/Spawners.cs:
```diff
@@ -193,7 +207,11 @@ private static void CCSpawnAI(ConCommandArgs args)
+            {
+                amount = 100;
+                Log.MessageNetworked($"Limited to 100, please dont spawn too many things at once.", args, LogLevel.MessageClientOnly);
+            }
```
Same cap arguments as for spawn_interactable and give_drone.

SChinchi commented on Code/StringFinder.cs:
```diff
@@ -359,6 +386,16 @@ public ItemIndex GetItemFromPartial(string name)
+        {
+            return GetPickupsFromPartial(name).DefaultIfEmpty(PickupIndex.none).First();
+        }
+
```
These are technically public API that others could use, even though nobody does at the moment. Doc strings would be nice.

SChinchi commented on Code/DT-Commands/Drones.cs:
```diff
@@ -0,0 +1,213 @@
+            Log.MessageNetworked(s, args, LogLevel.MessageClientOnly);
+        }
+
+        [ConCommand(commandName = "give_drone", flags = ConVarFlags.ExecuteOnServer, helpText = Lang.GIVEDRONE_HELP)]
```
Just a note on this that I had recently updated spawn_ai to spawn any body as an ally. Both that and give_drone can effectively spawn multiple allies but have some differences and we'll see whether both are needed.

@harbingerofme The TLDR is that `spawn_ai` allows any body instead of just drones, while `give_drone` allows to assign the drone's tier at spawning time. I was going to address this with a macro command that gives your drones the drone upgrade tier item. `give_drone` probably sounds more natural as the feature in `spawn_ai` might be missed by those who don't read all the details and just skim command names. And it pairs more naturally with the remove_drone commands. Thoughts?

SChinchi commented on Code/DT-Commands/CurrentRun.cs:
```diff
@@ -161,6 +162,40 @@ private static void CCNoEnemies(ConCommandArgs args)
+            else
+            {
+                SceneDirector.onPrePopulateSceneServer -= PreventInteractableSpawns;
+            }
```
There is an edge case with this approach. InfiniteTowerRun subscribes to this on its Start. If you call the command first and then start a simulacrum run, the run's event will be called second which overwrites your zeroed out credits. You'd then have to toggle the command again so your event runs last. In a similar vein we can also not account what mods do when they subscribe to this event. It seems to me the most watertight approach would be to ILHook PopulateScene and zero out the credits with an emit delegate just before the while loop.

SChinchi commented on Code/DT-Commands/CurrentRun.cs:
```diff
@@ -161,6 +162,40 @@ private static void CCNoEnemies(ConCommandArgs args)
+            if (noInteractables)
+            {
+                obj.interactableCredit = 0;
+                //obj.teleporterSpawnCard = null;
```
Remove if not needed.

SChinchi commented on Code/DT-Commands/CurrentRun.cs:
```diff
@@ -195,31 +230,84 @@ private static void CCKillAll(ConCommandArgs args)
+            else
+            {               
+                int count = 0;
+                int countV = 0;
```
Nitpick but better names here. I had no idea what these were doing until I saw them referenced. While we do use the name `count`/`amount` in a lot of commands, its purpose is more readily clear within the context of the command as we're really only counting one thing.

SChinchi reviewed (commented):
I'm submitting this review as general feedback, but there are some issues that should be addressed before merging. For the record the following classes have not been thoroughly reviewed as I lack the time and the mental energy at the moment. Expect a follow-up in a few days:
- Buffs
- Items
- PlayerCommands
- Hooks
- Lang
- StringFinder

WolfoIsBestWolf commented on Code/AutoCompletion/AutoCompleteManager.cs:
```diff
@@ -73,14 +73,15 @@ internal static void RegisterAutoCompleteCommands()
             parser.RegisterStaticVariable("body", BodyCatalog.allBodyPrefabBodyBodyComponents.Select(i => $"{(int)i.bodyIndex}|{i.name}|{StringFinder.GetLangInvar(i.baseNameToken)}"), 1);
             parser.RegisterStaticVariable("buff", BuffCatalog.buffDefs.Select(i => $"{(int)i.buffIndex}|{StringFinder.GetLangInvar(i.name)}"), 1);
-            parser.RegisterStaticVariable("droptable", ItemTierCatalog.allItemTierDefs.OrderBy(i => i.tier).Select(i => $"{(int)i.tier}|{i.name}"), 1);
+            parser.RegisterStaticVariable("tier", ItemTierCatalog.allItemTierDefs.OrderBy(i => i.tier).Select(i => $"{(int)i.tier}|{i.name}"), 1);
```
reverted

WolfoIsBestWolf commented on Code/DT-Commands/Miscellaneous.cs:
```diff
@@ -79,5 +80,53 @@ static IEnumerator InvokeRoutine(System.Action action, float delay)
+            }
+            Debug.Log(log);
+        }
+
```
Easier for me to get the mod guid from here than log and helps me double check if a modpack is *supposed* to be vanilla compatible.

WolfoIsBestWolf commented on Code/DT-Commands/PlayerCommands.cs:
```diff
@@ -606,5 +763,318 @@ internal static bool UpdateCurrentPlayerBody(out NetworkUser networkUser, out Ch
+        [ConCommand(commandName = "nocooldown", flags = ConVarFlags.ExecuteOnServer, helpText = Lang.NOCOOLDOWN_HELP)]
+        [ConCommand(commandName = "nocooldowns", flags = ConVarFlags.ExecuteOnServer, helpText = Lang.NOCOOLDOWN_HELP)]
+        [AutoComplete(Lang.TARGET_PLAYER_PINGED)]
+        public static void CC_Cooldown(ConCommandArgs args)
```
This would grant 0.001s cooldowns not 0.5. Without any ugly pink filter.

They'll look into it and else people can still use `give_item alien 5000`

Unless it's better to not ship the command with a known_issue with such a big mod.

WolfoIsBestWolf committed:
Cleanup 1

WolfoIsBestWolf committed:
Update PlayerCommands.cs

WolfoIsBestWolf committed:
'give_buff' negative amount now just calls 'remove_buff'

cleaner code this way.
also added same support for 'remove_buff'

WolfoIsBestWolf committed:
cleanse now a proper macro not rather duplicate code

calls the other commands now.

Apparently 'removed_all_buffs 0' is also just broken with ss2 at least with beta. Not sure what to do about that

WolfoIsBestWolf committed:
Update Buffs.cs

WolfoIsBestWolf committed:
Cleanup 2.0

WolfoIsBestWolf committed:
empty spaces / unused diffs again

WolfoIsBestWolf committed:
Fixed toggle_time not really working as it should if time_scale was already set

WolfoIsBestWolf committed:
Missed a type

WolfoIsBestWolf committed:
set_difficulty command

This should help if testing smth like how does it work on eclipse vs other difficulty.

Does mean the mod needs to depend on R2API_Difficulty.
-For modded difficulties
-For DifficultyDef to DifficutlyIndex (not invanilla I think)

WolfoIsBestWolf committed:
Made the set_ stats 1 stat to not fill up help command.

Do think this is still a good command to have.

Fixed Parse.BuffTarget 'pinged' returning wrong fail message if it fails.

WolfoIsBestWolf committed:
Removed 3 proposed commands due to redundancy.

dtdamage: Too plain

unlimited_junk: Too specific to need a toggle.

kill_all_minions: Seems a bit too specific, I use kill_all 1 all the time I wouldn't switch over to this.. But remove_all_drones is good.

WolfoIsBestWolf committed:
Small things

Changed 'create_pickup' [search] to an enum

Fixed an inconsistency with 'buddhaEnemy'

'charge_zone' default value 100%, instead of requiring you to give number.
(Seems unnecessary to ask for number)

Added dt style safety checks for 'loadout_set_skill_variant', 'skill', 'skin'

WolfoIsBestWolf committed:
Buffs.TryParse Pinged fail message fix for real

Did != instead of == oops

WolfoIsBestWolf committed:
Update CHANGELOG.md

WolfoIsBestWolf committed:
give_all_items {itemTier}

Devs got a command like this, good for like crafting testing

WolfoIsBestWolf committed:
'give_drone' -> `spawn_minion`

'remove_all_drones` -> `remove_all_minions`
'give_drone' -> Replaced with `spawn_minion`

Maybe more agreeable overlap?
And built upon already better code instead of more so duplicate.

I just like being able to get the minion I want with 1 argument not 4 but that might not be too important.
But other proposed commands are of similiar stature.

Replaces elite with droneTier and axes the TeamIndex argument.
If any other minion specific stuff might get added it could also just get added to this command yknow.

WolfoIsBestWolf committed:
Misc consistency of proposed changes

`toggle_hud` -> `hide_hud` consistency with hide_model and easier to type for me.

`true_kill all` now actually does that.

Lang for previous 2 commits.

WolfoIsBestWolf committed:
'spawn_portal' updated for DLC3

WolfoIsBestWolf committed:
Update Spawners.cs

SChinchi commented on Code/AutoCompletion/AutoCompleteManager.cs:
```diff
@@ -69,17 +70,27 @@ internal static void RegisterAutoCompleteCommands()
             var parser = new AutoCompleteParser();
             parser.RegisterStaticVariable("0", "0");
             parser.RegisterStaticVariable("1", "1");
+            parser.RegisterStaticVariable("count", "count");
+            parser.RegisterStaticVariable("droneTier", "droneTier");
+            parser.RegisterStaticVariable("duration", "duration");
+            parser.RegisterStaticVariable("skill_slot", "skill_slot");
+            parser.RegisterStaticVariable("skill_variant", "skill_variant");
```
The reason why `0` and `1` have been added as static variables is so that in the command args we can do `{some_bool (0|1)}` instead of `{some_bool ('0'|'1')}`. They're basically there for convenience. When you add "count" there, in `give_item {item} [count]` the count will appear as an actual autocomplete option when typing the second argument, which we don't want.

SChinchi commented on Code/AutoCompletion/AutoCompleteManager.cs:
```diff
@@ -69,17 +70,27 @@ internal static void RegisterAutoCompleteCommands()
             parser.RegisterStaticVariable("buff", BuffCatalog.buffDefs.Select(i => $"{(int)i.buffIndex}|{StringFinder.GetLangInvar(i.name)}"), 1);
-            parser.RegisterStaticVariable("tier", ItemTierCatalog.allItemTierDefs.OrderBy(i => i.tier).Select(i => $"{(int)i.tier}|{i.name}"), 1);
+            parser.RegisterStaticVariable("droptable", ItemTierCatalog.allItemTierDefs.OrderBy(i => i.tier).Select(i => $"{(int)i.tier}|{i.name}"), 1);
+            parser.RegisterStaticVariable("itemTier", ItemTierCatalog.allItemTierDefs.OrderBy(i => i.tier).Select(i => $"{(int)i.tier}|{i.name}"), 1);
```
Can we rename this to `itemtier`? Mainly because we have already established with `list_itemtier` the expectation of what that should be called. But also for the consistency of snake case. It would be nice we if could also rename `droneTier` to either `drone_tier` or `dronetier`. Whichever you prefer, though the former looks better to me.

SChinchi commented on Code/DT-Commands/Buffs.cs:
```diff
@@ -90,7 +90,8 @@ private static void CCGiveBuff(ConCommandArgs args)
             if (iCount < 0)
             {
-                Log.MessageNetworked(String.Format(Lang.NEGATIVE_ARG, "count"), args, LogLevel.MessageClientOnly);
+                args.userArgs[1] = (-iCount).ToString();
+                CCRemoveBuff(args);
```
`give_buff` is technically not syntactically the opposite of `remove_buff`, which is why negative stacks were not supported from the beginning. `give_buff buff 1 4` intends to add one stack for 4 seconds. -1 stacks here wouldn't make sense. In fact, `remove_buff` for timed stacks always removes the oldest timed stack.

SChinchi commented on Code/DT-Commands/Buffs.cs:
```diff
@@ -147,7 +148,7 @@ private static void CCGiveBuff(ConCommandArgs args)
                     body.AddTimedBuff(buff, duration);
                 }
-                Log.MessageNetworked($"Gave {iCount} {name} to {target.name} for {duration} seconds", args);
+                Log.MessageNetworked($"Gave {iCount} {name} to {target.name} for <color=#53E9FF>{duration} seconds</color>", args);
```
I think we should avoid colors in logs as much as possible. The main reason I've used green/red for enabled/disabled is because for toggle commands it's very easy to not notice if you're toggling it to the wrong value.

SChinchi commented on Code/DT-Commands/Buffs.cs:
```diff
@@ -175,7 +176,8 @@ private static void CCRemoveBuff(ConCommandArgs args)
             }
             if (iCount < 0)
             {
-                Log.MessageNetworked(String.Format(Lang.NEGATIVE_ARG, "count"), args, LogLevel.MessageClientOnly);
+                args.userArgs[1] = (-iCount).ToString();
+                CCGiveBuff(args);
```
Same argument as above. Negative stacks for `remove_buff` is not synonymous with positive stacks of `give_buff`. remove_buff's 3rd argument is supposed to be a bool for whether the buff type is timed or not. give_buff on the other hand expects an explicit integer for the duration of the timed buff. And yes, I'm aware that you can technically pass any non-zero value for bool and it'll count as true, but we'll getting into some implicit conversions here that are not as straightforward as just `give_item` -> `remove_item` for negative stacks.

SChinchi commented on Code/DT-Commands/Macros.cs:
```diff
@@ -35,15 +35,40 @@ private static void LateGame(ConCommandArgs args)
+        [ConCommand(commandName = "dtpeace", flags = ConVarFlags.ExecuteOnServer, helpText = Lang.MACRO_DTPEACE_HELP)]
+        private static void Peace(ConCommandArgs args)
+        {
+            Invoke(args.sender, "kill_all", "2", "1");
+            Invoke(args.sender, "kill_all", "4", "1");
```
If we make the `kill_all enemies` suggested change, we can also apply here.

SChinchi commented on Code/DT-Commands/Spawners.cs:
```diff
@@ -378,11 +435,20 @@ internal static void InitPortals()
             portals.Add("artifact", Addressables.LoadAssetAsync<GameObject>("RoR2/Base/PortalArtifactworld/PortalArtifactworld.prefab").WaitForCompletion());
             portals.Add("blue", Addressables.LoadAssetAsync<GameObject>("RoR2/Base/PortalShop/PortalShop.prefab").WaitForCompletion());
             portals.Add("celestial", Addressables.LoadAssetAsync<GameObject>("RoR2/Base/PortalMS/PortalMS.prefab").WaitForCompletion());
-            portals.Add("deepvoid", Addressables.LoadAssetAsync<GameObject>("RoR2/DLC1/DeepVoidPortal/DeepVoidPortal.prefab").WaitForCompletion());
             portals.Add("gold", Addressables.LoadAssetAsync<GameObject>("RoR2/Base/PortalGoldshores/PortalGoldshores.prefab").WaitForCompletion());
-            portals.Add("green", Addressables.LoadAssetAsync<GameObject>("RoR2/DLC2/PortalColossus.prefab").WaitForCompletion());
             portals.Add("null", Addressables.LoadAssetAsync<GameObject>("RoR2/Base/PortalArena/PortalArena.prefab").WaitForCompletion());
+            //DLC1
             portals.Add("void", Addressables.LoadAssetAsync<GameObject>("RoR2/DLC1/PortalVoid/PortalVoid.prefab").WaitForCompletion());
+            portals.Add("deepvoid", Addressables.LoadAssetAsync<GameObject>("RoR2/DLC1/DeepVoidPortal/DeepVoidPortal.prefab").WaitForCompletion());
+            portals.Add("simulacrum", Addressables.LoadAssetAsync<GameObject>("RoR2/DLC1/GameModes/InfiniteTowerRun/PortalInfiniteTower.prefab").WaitForCompletion());
+            //DLC2
+            portals.Add("green", Addressables.LoadAssetAsync<GameObject>("RoR2/DLC2/PortalColossus.prefab").WaitForCompletion());
+            portals.Add("destination", Addressables.LoadAssetAsync<GameObject>("RoR2/DLC2/PM DestinationPortal.prefab").WaitForCompletion());
+            //DLC3
+            portals.Add("encrypted", Addressables.LoadAssetAsync<GameObject>("RoR2/DLC3/HardwareProgPortal.prefab").WaitForCompletion());
+            portals.Add("decrypted", Addressables.LoadAssetAsync<GameObject>("RoR2/DLC3/SolusWebPortal.prefab").WaitForCompletion());
+            portals.Add("virtual", Addressables.LoadAssetAsync<GameObject>("RoR2/DLC3/CompExchangePortal.prefab").WaitForCompletion());
+            portals.Add("mainline", Addressables.LoadAssetAsync<GameObject>("RoR2/DLC3/EyePortal/EyePortal.prefab").WaitForCompletion());
```
There are a few portals here, like simulacrum and destination, that don't do anything fancy under any circumstances; they just send you to whatever is the next stage. Which is why they had been omitted.

A bit of a rant here, but the reason `spawn_portal` was created in the first place was because the null/deepvoid portals used to be part of `add_portal` but because there were no orbs, that would spawn directly on the field. So I made `add_portal` to be purely for the TP orbs and `spawn_portal` for spawning directly. But to tell you the truth, I don't know who would even use this command and it's annoying in terms of maintaining with each new DLC. I'll check with the team to see if there's any point keeping it around.

SChinchi commented on Code/DebugToolkit.cs:
```diff
@@ -73,6 +73,7 @@ private void LogBuildInfo()
         private void Start()
         {
             var _ = StringFinder.Instance;
+            RoR2Application.onLoad += AutoCompleteManager.RegisterAutoCompleteCommands;
```
Why was this moved?

SChinchi commented on Code/Hooks.cs:
```diff
@@ -156,18 +236,10 @@ private static void SetGodMode(On.RoR2.CharacterMaster.orig_Awake orig, Characte
             if (NetworkServer.active)
             {
                 self.godMode |= self.playerCharacterMasterController && god;
+                self.godMode |= self.teamIndex == TeamIndex.Monster && godMonsters;
```
`self.TeamIndex != TeamIndex.Player && godMonsters` makes more sense to me.

SChinchi commented on Code/Hooks.cs:
```diff
@@ -93,12 +94,91 @@ public static void InitializeHooks()
+
+            On.RoR2.ConCommandArgExtensions.TryGetArgBodyIndex += AllowBodyIndexToBeUsedInsteadOfBodyName;
+            On.RoR2.UserProfile.CCLoadoutSetSkillVariant += LoadoutSetSkillVariant_AcceptSELF;
+            On.RoR2.Run.CCRunSetStagesCleared += RunSetStagesCleared_SetLoopCountToo;
+        }
+
+        private static void RunSetStagesCleared_SetLoopCountToo(On.RoR2.Run.orig_CCRunSetStagesCleared orig, ConCommandArgs args)
+        {
+            orig(args);
+            if (Run.instance)
+            {
+                Run.instance.Network_loopClearCount = Run.instance.NetworkstageClearCount / 5;
+            }
+        }
```
This modifies vanilla behavior and does not belong in DebugToolkit. It is also not networked.

SChinchi commented on Code/DT-Commands/Miscellaneous.cs:
```diff
@@ -79,5 +80,36 @@ static IEnumerator InvokeRoutine(System.Action action, float delay)
+                log.Append("Mods tagged as RequiredByAll\n\n");
+                foreach (var a in NetworkModCompatibilityHelper.networkModList)
+                {
+                    log.Append(a.ToString());
+                    log.Append("\n");
```
Substitute with `log.AppendLine(a)`;

SChinchi commented on Code/DT-Commands/Miscellaneous.cs:
```diff
@@ -79,5 +80,36 @@ static IEnumerator InvokeRoutine(System.Action action, float delay)
+                    log.Append("\n");
+                }
+            }
+            Log.Message(log.ToString());
```
`log.ToString().TrimEnd('\n')` to get rid of any final newlines.

SChinchi commented on Code/DT-Commands/Money.cs:
```diff
@@ -20,7 +22,7 @@ private static void CCGiveLunar(ConCommandArgs args)
             }
             if (args.sender == null)
             {
-                Log.Message("Can't modify Lunar coins of other users directly.", LogLevel.MessageClientOnly);
+                Log.Message("Can't modify Lunar Coins/Void Markers of other users directly.", LogLevel.MessageClientOnly);
```
You can figure out whether to say "Lunar Coins" or "Void Markers" based on the name of the command.

SChinchi commented on Code/DT-Commands/PlayerCommands.cs:
```diff
@@ -107,6 +196,93 @@ private static void CCCursorTeleport(ConCommandArgs args)
+                {
+                    powerPedestal.SetComplete(true);
+                }
+                Debug.Log("Autocompleting Sentry Terminals");
```
Change this to the `Log` class every where in this command.

SChinchi commented on Code/DT-Commands/PlayerCommands.cs:
```diff
@@ -107,6 +196,93 @@ private static void CCCursorTeleport(ConCommandArgs args)
+            Vector3 newPosition = Vector3.zero;
+            string destination = string.Empty;
+
+            if (TeleporterInteraction.instance)
+            {
+                destination = "Teleporter";
+                newPosition = TeleporterInteraction.instance.transform.position;
+                newPosition += new Vector3(0, 1, 0);
+            }
+            else if (BossGroup.GetTotalBossCount() > 0)
+            {
+                for (int i = 0; i < CharacterBody.instancesList.Count; i++)
+                {
+                    if (CharacterBody.instancesList[i].isBoss)
+                    {
+                        newPosition = CharacterBody.instancesList[i].corePosition;
+                        destination = CharacterBody.instancesList[i].name;
+                        break;
+                    }
+                }
+            }
```
I'm happy to support the final boss explicit locations, but this one not so much. Wouldn't `BossGroup.GetTotalBossCount` also count things like umbras and AWU?

SChinchi commented on Code/DT-Commands/PlayerCommands.cs:
```diff
@@ -107,6 +196,93 @@ private static void CCCursorTeleport(ConCommandArgs args)
+                destination = "Voidling Arena";
+                newPosition = new Vector3(-105f, 0.2f, 92f);
+            }
+            if (stage == "conduitcanyon")
```
This one has a very different result from all the other ones, I don't think it belongs here.

SChinchi commented on Code/DT-Commands/CurrentRun.cs:
```diff
@@ -195,15 +230,19 @@ private static void CCKillAll(ConCommandArgs args)
+                DebugToolkit.InvokeCMD(args.sender, "kill_all", ((int)TeamIndex.Void).ToString());
+                return;
+            }
+
```
It would be better to modify the team argument to `[team (team|'enemies'): Monster]` and do something like

```cs
TeamIndex[] targetTeams = new[] { TeamIndex.Monster };
if (arg.Count > 0 || args[0] != Lang.DEFAULT_VALUE)
{
    if (arg == 'enemies')
    {
        targetTeams = allNotPlayerTeams;
    }
    else if (TryParse(arg, out team))
    {
        targetTeams = new[] { team };
    }
    else
    {
        Log("Team not found");
        return;
    }
}

foreach (var teamIndex in targetTeams)
{
    DoTheThing();
}
```

The log can then just be "killed X of team Y" or "killed X total enemies". `Lang.KILLALL_HELP` could also do with an edit here to explain that enemies refers to all non-player teams.

SChinchi commented on Code/DT-Commands/PlayerCommands.cs:
```diff
@@ -107,6 +196,93 @@ private static void CCCursorTeleport(ConCommandArgs args)
+                }
+                Debug.Log("Autocompleting Sentry Terminals");
+            }
+            if (newPosition == Vector3.zero)
```
Would be nice to also include `solusweb` here.

SChinchi reviewed (changes requested):
Second round of review. There are still some big sections that I haven't covered because there are widespread changes related to them.

For `create_pickup` I had been thinking for some time now changing the signature to `create_pickup {pickup} [pickup_type: Permanent]`. It immensely simplifies things as there are no more search options, etc. However, it does mean it backwards breaking API change, so I need to run it with the rest of the team. With the release of AC the usefulness of using the pickup catalog has increased further, which is why I am reconsidering this change. This also means I can finally implement `set_pickup {pickup}` which rerolls the pickup for the pinged/nearest ShopTerminalBehavior/DroneVendorTerminalBehavior/whatever. Depending on what decision is taken, it might be simpler to just revert any changes you've made to `create_pickup` so I can do my thing.

The other is reverting any changes/macros related to loadout_skill/loadout_skin commands. We shouldn't modify how vanilla code behaves, even if it's just for our added convenience. Unlocking cheat commands is one thing, but changing how the arguments to a command work is another. Furthermore, what I had been thinking about for some time now is reimplementing `loadout_set_skill_variant` (and the skin one) in our own custom way, because the vanilla one lacks some features. The command executes on server, but only updates the profile data locally, which means clients don't get full use out of it. And if I get to write them my own way, I can reorder the arguments as `{slot} {skill} [body]` which would easily allow the body to use the self as default. Along with supporting `pinged` for monsters (mainly for player AI but still) and some other things.
